open-apparel-registry/open-apparel-registry
# Add API limits management command

## Overview

Creates a management command to:

* Count API requests
* Create/update ApiBlock records
* Send notification emails about changes to ApiBlock records

Connects #1139 

## Demo

<img width="457" alt="Screen Shot 2020-10-28 at 9 59 00 AM" src="https://user-images.githubusercontent.com/21046714/97445969-3bdf8500-1904-11eb-80e9-0181efe9667d.png">
<img width="595" alt="Screen Shot 2020-10-28 at 9 58 49 AM" src="https://user-images.githubusercontent.com/21046714/97445972-3da94880-1904-11eb-9b72-d53abcafde4d.png">


## Testing Instructions

* Checkout this branch
* Run `./scripts/update`
* Set a contributor's ApiLimit.monthly_limit to 0 using the admin dashboard.
* Run `./scripts/manage check_api_limits`. Nothing should happen.
* Set the contributor's ApiLimit.monthly_limit to 10.
* Use `./scripts/manage shell_plus` to create a RequestLog for the contributor, setting the date within the current month the response code in the 200s.
* Run `./scripts/manage check_api_limits`. Nothing should happen.
* Add eight more RequestLogs for the contributor (as above), for a total of nine.
* Run `./scripts/manage check_api_limits`. You should see an email printed to the console. Check the ContributorNotifications for the user; the api_limit_warning_sent_on should equal the current date.
* Run `./scripts/manage check_api_limits` again. Nothing should happen.
* Add one more RequestLog for the contributor. 
* Run `./scripts/manage check_api_limits`. You should see an email printed to the console. Check the ContributorNotifications for the user; the api_limit_exceeded_sent_on should equal the current date. 
* There should be an active ApiBlock set for the contributor with an end date set to the end of the current month. 
* Run `./scripts/manage check_api_limits`. Nothing should happen.
* Set the grace limit for the ApiBlock to 12 and set active to False. 
* Run `./scripts/manage check_api_limits`. Nothing should happen.
* Create two more RequestLogs for the contributor.
* Run `./scripts/manage check_api_limits`. Nothing should happen. You should see an email printed to the console. Check the ContributorNotifications for the user; the api_grace_limit_exceeded_sent_on should equal the current date. The ApiBlock should be active again.
* Run `./scripts/manage check_api_limits`. Nothing should happen.
* Run `./scripts/test`. All tests should pass.

## Checklist
